br/pkg/lightning/errormanager/errormanager.go

```diff
@@ -642,11 +651,116 @@ func (em *ErrorManager) ReplaceConflictKeys(
 				}
 			}
 		}
-		if err := rawKeyRows.Err(); err != nil {
+		if err := indexKvRows.Err(); err != nil {
+			return errors.Trace(err)
+		}
+
+		// check data KV
+		dataKvRows, err := em.db.QueryContext(
+			gCtx, fmt.Sprintf(selectDataConflictKeysReplace, em.schemaEscaped),
+			tableName)
+		if err != nil {
 			return errors.Trace(err)
 		}
+		defer dataKvRows.Close()
+
+		var previousRawKey, latestValue []byte
+		var mustKeepKvPairs *kv.Pairs
+
+		for dataKvRows.Next() {
+			var rawKey, rawValue, rawHandle []byte
+			if err := dataKvRows.Scan(&rawKey, &rawValue, &rawHandle); err != nil {
+				return errors.Trace(err)
+			}
+			em.logger.Debug("got group raw_key, raw_value, raw_handle from table",
+				logutil.Key("raw_key", rawKey),
+				zap.Binary("raw_value", rawValue),
+				zap.Binary("raw_handle", rawHandle))
+
+			if !bytes.Equal(rawKey, previousRawKey) {
+				previousRawKey = rawKey
+				// get the latest value of rawKey from downstream TiDB
+				latestValue, err = fnGetLatest(gCtx, rawKey)
+				if err != nil && !tikverr.IsErrNotFound(err) {
+					return errors.Trace(err)
+				}
+				if latestValue != nil {
+					handle, err := tablecodec.DecodeRowKey(rawKey)
+					if err != nil {
+						return errors.Trace(err)
+					}
+					decodedData, _, err := tables.DecodeRawRowData(encoder.SessionCtx,
+						tbl.Meta(), handle, tbl.Cols(), latestValue)
+					if err != nil {
+						return errors.Trace(err)
+					}
+					_, err = encoder.Table.AddRecord(encoder.SessionCtx, decodedData)
+					if err != nil {
+						return errors.Trace(err)
+					}
+					// calculate the new mustKeepKvPairs corresponding to the new rawKey
+					// find out all the KV pairs that are contained in the data KV
+					mustKeepKvPairs = encoder.SessionCtx.TakeKvPairs()
+				}
+			}
 
-		// TODO: check data KV
+			// if the latest value of rawKey equals to rawValue, that means this data KV is maintained in downstream TiDB
+			// if not, that means this data KV has been deleted due to overwritten index KV
+			if bytes.Equal(rawValue, latestValue) {
```

Comment:
we only need to decode `latestValue` when `!bytes.Equal(rawKey, previousRawKey)`, so if we put this check before we decode `latestValue`, we need to check `!bytes.Equal(rawKey, previousRawKey)` again to determine whether we need to decode `latestValue`, which seems unnecessary

Reply:
no need to check twice, like this:
```go
			if bytes.Equal(rawKey, previousRawKey) {
				continue
			}
			previousRawKey = rawKey
			// get the latest value of rawKey from downstream TiDB
			latestValue, err = fnGetLatest(gCtx, rawKey)
			if err != nil && !tikverr.IsErrNotFound(err) {
				return errors.Trace(err)
			}

			// TODO: check data KV				// if the latest value of rawKey equals to rawValue, that means this data KV is maintained in downstream TiDB
			// if not, that means this data KV has been deleted due to overwritten index KV
			if bytes.Equal(rawValue, latestValue) {
				continue
			}
			if latestValue != nil {
				handle, err := tablecodec.DecodeRowKey(rawKey)
				if err != nil {
					return errors.Trace(err)
				}
				decodedData, _, err := tables.DecodeRawRowData(encoder.SessionCtx,
					tbl.Meta(), handle, tbl.Cols(), latestValue)
				if err != nil {
					return errors.Trace(err)
				}
				_, err = encoder.Table.AddRecord(encoder.SessionCtx, decodedData)
				if err != nil {
					return errors.Trace(err)
				}
				// calculate the new mustKeepKvPairs corresponding to the new rawKey
				// find out all the KV pairs that are contained in the data KV
				mustKeepKvPairs = encoder.SessionCtx.TakeKvPairs()
			}
```

Reply:
we cannot directly write this 
```go
if bytes.Equal(rawKey, previousRawKey) {
	continue
}
```
because if rawKey equals to previousRawKey, we still need to conduct the code from line 707 to line 763